### PR TITLE
Feature/full core usage

### PR DESCRIPTION
**Pull request by BabysbreathJJ**
#109

### CODE REVIEW

File: src/soa/SessionLauncher/Impls/JobMonitorEntry/AzureBatch/AzureBatchJobMonitor.cs [hunk 1]
```diff
@@ -114,12 +146,21 @@ public async Task StartAsync()
...
+                "[AzureBatchJobMonitorEntry] Enters QueryTaskInfo method.");
             bool shouldExit = false;
             this.pullJobGap = PullJobMinGap;
             JobState state = JobState.Active;
             Session.Data.JobState currentJobState = Session.Data.JobState.Configuring;
             var pool = this.batchClient.PoolOperations.GetPool(AzureBatchConfiguration.BatchPoolName);
+            string skuName = pool.VirtualMachineSize;
+            TraceHelper.TraceEvent(this.sessionid, TraceEventType.Information, "[AzureBatchJobMonitor] VMSize in pool is {0}",
+                skuName);
+            SKU targetSku = Array.Find(this.skus, sku => sku.name.ToLower().Equals(skuName));
```
Comment (amat27) [inline src/soa/SessionLauncher/Impls/JobMonitorEntry/AzureBatch/AzureBatchJobMonitor.cs]: `sku.name.Equals(skuName, StringComparison.OrdinalIgnoreCase)`

File: src/soa/SessionLauncher/Impls/JobMonitorEntry/AzureBatch/AzureBatchJobMonitor.cs [hunk 2]
```diff
@@ -73,6 +76,34 @@ internal class AzureBatchJobMonitor : IDisposable
 
         private Telepathy.Session.Data.JobState previousJobState = Session.Data.JobState.Configuring;
 
+        class SKU
```
Comment (amat27) [inline src/soa/SessionLauncher/Impls/JobMonitorEntry/AzureBatch/AzureBatchJobMonitor.cs]: Avoid using nesting class. Move `SKU` to separate file, change its fields to properties, and mark it internal.

File: src/soa/SessionLauncher/Impls/JobMonitorEntry/AzureBatch/AzureBatchJobMonitor.cs [hunk 3]
```diff
@@ -73,6 +76,34 @@ internal class AzureBatchJobMonitor : IDisposable
...
+        {
+            public string name;
+            public string vCPUs;
```
Comment (amat27) [inline src/soa/SessionLauncher/Impls/JobMonitorEntry/AzureBatch/AzureBatchJobMonitor.cs]: Why we don't parse `vCPUs` to `int` here?

File: src/soa/SessionLauncher/Impls/JobMonitorEntry/AzureBatch/AzureBatchJobMonitor.cs [hunk 4]
```diff
@@ -254,27 +292,36 @@ private async Task<List<TaskInfo>> GetTaskStateChangeAsync(List<ComputeNode> nod
                 {
                     TraceHelper.TraceEvent(this.sessionid, TraceEventType.Information, "[AzureBatchJobMonitor] task {0} state changed to {1}, at date time = {2}\n", task.Id, task.State, task.StateTransitionTime);
                     TaskState state = task.State.Value;
+                    DateTime stateTransitionTime = (DateTime)task.StateTransitionTime;
```
Comment (amat27) [inline src/soa/SessionLauncher/Impls/JobMonitorEntry/AzureBatch/AzureBatchJobMonitor.cs]: `task.StateTransitionTime.Value`

File: src/soa/SessionLauncher/Impls/JobMonitorEntry/AzureBatch/AzureBatchJobMonitor.cs [hunk 5]
```diff
@@ -254,27 +292,36 @@ private async Task<List<TaskInfo>> GetTaskStateChangeAsync(List<ComputeNode> nod
...
+                            Id = task.Id,
+                            State = TaskStateConverter.FromAzureBatchTaskState(task.State.Value)
+                        };
                         results.Add(info);
                     }
+
+                    if (DateTime.Compare(this.lastChangeTime, stateTransitionTime) < 1)
+                    {
+                        this.lastChangeTime = stateTransitionTime;
+                    }
```
Comment (amat27) [inline src/soa/SessionLauncher/Impls/JobMonitorEntry/AzureBatch/AzureBatchJobMonitor.cs]: If an exception happens in the middle of this method, will we miss task events again?

File: src/soa/SessionLauncher/SKUs.json [hunk 6]
```diff
@@ -0,0 +1 @@
...128-32ms","vCPUs":"128"},{"name":"Standard_M128-64ms","vCPUs":"128"},{"name":"Standard_M128ms","vCPUs":"128"},{"name":"Standard_M128s","vCPUs":"128"},{"name":"Standard_M64","vCPUs":"64"},{"name":"Standard_M64m","vCPUs":"64"},{"name":"Standard_M128","vCPUs":"128"},{"name":"Standard_M128m","vCPUs":"128"},{"name":"Standard_A8","vCPUs":"8"},{"name":"Standard_A9","vCPUs":"16"},{"name":"Standard_A10","vCPUs":"8"},{"name":"Standard_A11","vCPUs":"16"},{"name":"Standard_ND40rs_v2","vCPUs":"40"},{"name":"Standard_G1","vCPUs":"2"},{"name":"Standard_G2","vCPUs":"4"},{"name":"Standard_G3","vCPUs":"8"},{"name":"Standard_G4","vCPUs":"16"},{"name":"Standard_G5","vCPUs":"32"},{"name":"Standard_GS1","vCPUs":"2"},{"name":"Standard_GS2","vCPUs":"4"},{"name":"Standard_GS3","vCPUs":"8"},{"name":"Standard_GS4","vCPUs":"16"},{"name":"Standard_GS4-4","vCPUs":"16"},{"name":"Standard_GS4-8","vCPUs":"16"},{"name":"Standard_GS5","vCPUs":"32"},{"name":"Standard_GS5-8","vCPUs":"32"},{"name":"Standard_GS5-16","vCPUs":"32"},{"name":"Standard_L4s","vCPUs":"4"},{"name":"Standard_L8s","vCPUs":"8"},{"name":"Standard_L16s","vCPUs":"16"},{"name":"Standard_L32s","vCPUs":"32"},{"name":"Standard_HB120rs_v2","vCPUs":"120"}]
```
Comment (amat27) [inline src/soa/SessionLauncher/SKUs.json]: This is really nice. Can we format it for easy reading and editing?

File: src/soa/SessionLauncher/Impls/JobMonitorEntry/AzureBatch/AzureBatchJobMonitor.cs (hunk 5 above)
Comment (BabysbreathJJ) [reply src/soa/SessionLauncher/Impls/JobMonitorEntry/AzureBatch/AzureBatchJobMonitor.cs]: Add a temp variable to record the last transition time, and change the lastChangeTime in the end.

**Pull request merged**